### PR TITLE
[DeepSeek V4.1] Add DeepSelect AOT kernel.

### PR DESCRIPTION
**Pull request by yuyu5333**
## Motivation

This PR replaces the runtime dependency on the external DeepSelect package with an SM90-only AOT kernel in `sgl-kernel` .

It also restores fused candidate block-score reduction and mask publication,
preserving the optimized candidate-selection pipeline introduced in #39421.

## Modifications

For SM90:

``` bash
# before pr
select_candidate_blocks()
  -> _candidate_block_topk()
     -> PyTorch block-score reduction
        F.pad()
        -> unflatten()
        -> amax()
        -> masked_fill()
     -> torch.topk()
  -> PyTorch mask publication
     zeros()
     -> scatter()
     -> repeat_interleave()

# after pr
select_candidate_blocks()
  -> _candidate_block_topk()
     -> if CUDA SM90 and the AOT operator is available:
          fused Triton candidate_block_scores()
          -> sgl_kernel.deepselect_topk_fp32()
             -> torch.ops.sgl_kernel.deepselect_topk_fp32()
                -> prebuilt deepselect_ops.so
        else:
          PyTorch block-score reduction
          -> torch.topk()
  -> if the fused SM90 path was used:
       fused Triton publish_candidate_block_mask()
     else:
       PyTorch mask publication
```

The SM90 fast path is:

```bash
fused Triton block-score reduction
  -> DeepSelect FP32 AOT Top-K
  -> fused Triton token-mask publication
  -> downstream paged Top-K
```

This PR only adds the DeepSelect FP32 AOT path for NVIDIA Hopper SM90. 

###  Future work for SM100/SM103

This PR does not change the SM100/SM103 runtime path. These architectures currently keep their existing DSV4 JIT Top-K implementations:

``` bash
DeepSeek V4.1 runtime
  -> sglang.kernels.ops.attention.dsv4.topk.py
     -> _jit_topk_*_module()
        -> load_jit(...)
           -> compile CUDA source at runtime
              deepseek_v4/topk_v1.cuh
              deepseek_v4/topk_v2.cuh
              deepseek_v4/topk_bf16_small.cuh
        -> module.topk_transform(...)
```

A follow-up can migrate the SM100/SM103 implementations to architecture-specific AOT operators:

``` bash
DeepSeek V4.1 runtime
  -> sglang.kernels.ops.attention.dsv4.topk.py
     -> torch.ops.sgl_kernel.<aot_topk_op>(...)
        -> prebuilt sgl-kernel shared library
           -> architecture-specific CUDA kernel
```

The intended transition is:

``` bash
Current:
Python -> load_jit() -> runtime CUDA compilation -> JIT module

Future:
Python -> torch.ops.sgl_kernel.* -> prebuilt AOT shared library
```

## Speed Tests and Profiling

The complete candidate path was benchmarked on NVIDIA H20 using:

- FP32 candidate logits.
- `topk_blocks=2048`.
- 20 warmup iterations.
- 100 measured iterations.
- Median latency.
- L2 cache flushing before every measured iteration.
- Two complete benchmark runs; the tables report the mean of the two medians.

The measured region includes:

```text
candidate block-score reduction
  -> candidate Top-K
  -> candidate-mask publication
  -> downstream paged Top-K
```

### Final Top-K 512

| Rows | Sequence length | PyTorch | Fused AOT | Speedup |
|---:|---:|---:|---:|---:|
| 1 | 128K | 153.264 us | 44.112 us | 3.47x |
| 1 | 256K | 179.744 us | 63.712 us | 2.82x |
| 1 | 512K | 179.456 us | 73.888 us | 2.43x |
| 1 | 1M | 188.976 us | 96.624 us | 1.96x |
| 6 | 128K | 166.032 us | 44.064 us | 3.77x |
| 6 | 256K | 180.048 us | 59.648 us | 3.02x |
| 6 | 512K | 203.168 us | 73.168 us | 2.78x |
| 6 | 1M | 277.936 us | 104.672 us | 2.66x |

### Final Top-K 2048

| Rows | Sequence length | PyTorch | Fused AOT | Speedup |
|---:|---:|---:|---:|---:|
| 1 | 128K | 154.304 us | 45.056 us | 3.42x |
| 1 | 256K | 176.864 us | 64.720 us | 2.73x |
| 1 | 512K | 179.872 us | 75.168 us | 2.39x |
| 1 | 1M | 187.600 us | 98.096 us | 1.91x |
| 6 | 128K | 167.328 us | 45.168 us | 3.70x |
| 6 | 256K | 180.592 us | 60.656 us | 2.98x |
| 6 | 512K | 204.368 us | 74.256 us | 2.75x |
| 6 | 1M | 279.296 us | 106.032 us | 2.63x |

Across all 16 shapes:

- Speedup range: `1.91x–3.77x`.
- Geometric-mean speedup: `2.79x`.
- Best result: `3.77x` for 6 rows, 128K sequence length, and final Top-K 512.

The AOT Top-K kernel was also compared directly with the external DeepSelect implementation at width 131072 and Top-K 2048. The latency difference was approximately 0.3% to 0.4%, indicating that moving the kernel into `sgl-kernel` does not introduce a kernel-level performance regression.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #35724437450](https://github.com/sgl-project/sglang/actions/runs/35724437450)
Latest PR Test (Extra): :x: [Run #35724437228](https://github.com/sgl-project/sglang/actions/runs/35724437228)
Latest PR Test (AMD ROCm 10): :x: [Run #35724437521](https://github.com/sgl-project/sglang/actions/runs/35724437521)

### CODE REVIEW

File: python/sglang/kernels/aot/csrc/deepselect/3rdparty/kerutils/include/kerutils/device/cuda/sm80/intrinsics.cuh [hunk 1]
```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include "kerutils/device/cuda/common.h"
+
+namespace kerutils {
+
+// cp.async.cg (cache global) with prefetch and predicate (https://docs.nvidia.com/cuda/parallel-thread-execution/#data-movement-and-conversion-instructions-cp-async)
+template<PrefetchSize PREFETCH_SIZE=PrefetchSize::B128>
+CUTE_DEVICE
+void cp_async_cacheglobal(const void* src, void* dst, bool pred=true) {
```
Comment (DarkSharpness) [inline python/sglang/kernels/aot/csrc/deepselect/3rdparty/kerutils/include/kerutils/device/cuda/sm80/intrinsics.cuh:10]: Is this really used? I'm not sure whether we should prune the dead code in dependencies cc @BBuf

File: python/sglang/kernels/aot/python/sgl_kernel/__init__.py [hunk 2]
```diff
@@ -135,6 +136,12 @@
         from sgl_kernel.elementwise import gelu_quick
         from sgl_kernel.top_k import deepseek_v4_topk_transform_512
 
+    if torch.version.cuda is not None and _get_compute_capability() == 90:
+        try:
+            from sgl_kernel.deepselect import deepselect_topk_fp32
+        except ImportError:
+            pass
```
Comment (DarkSharpness) [inline python/sglang/kernels/aot/python/sgl_kernel/__init__.py]: Why restricted to sm90? Can we enable this for both sm90 and sm100?

File: python/sglang/kernels/aot/python/sgl_kernel/deepselect.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+def _aligned_empty(rows: int, cols: int, alignment_bytes: int, dtype, device):
+    alignment = alignment_bytes // dtype.itemsize
+    stride = (cols + alignment - 1) // alignment * alignment
+    return torch.empty((rows, stride), dtype=dtype, device=device)[:, :cols]
+
+
+def deepselect_topk_fp32(
+    input: torch.Tensor, topk: int
+) -> tuple[torch.Tensor, torch.Tensor]:
```
Comment (DarkSharpness) [inline python/sglang/kernels/aot/python/sgl_kernel/deepselect.py]: Same. We should allow all input type here (bf16, fp32). Also, please keep the same interface with the official version. The `topk_lengths` is needed for varlen decode case.

File: python/sglang/srt/layers/attention/dsv4/candidate_indexer.py
Comment (DarkSharpness) [inline python/sglang/srt/layers/attention/dsv4/candidate_indexer.py:1]: let's keep all the runtime integration in another PR. For this PR, only update the AOT-kernel related part

File: python/sglang/kernels/aot/python/sgl_kernel/deepselect.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    """Return unsorted FP32 values and int32 indices for each input row."""
+    if input.dtype != torch.float32 or input.dim() != 2:
+        raise ValueError("input must be a 2D float32 tensor")
+    if not input.is_cuda:
+        raise ValueError("input must be a CUDA tensor")
+    if not 0 < topk <= min(4096, input.shape[1]):
+        raise ValueError("topk must be in [1, min(4096, input.shape[1])]")
```
Comment (DarkSharpness) [inline python/sglang/kernels/aot/python/sgl_kernel/deepselect.py]: If the checks already exist in C++, please remove the redundant parts in Python.

File: python/sglang/kernels/aot/csrc/deepselect/cuda_kernels/v3_fp32/instantiations/value_fp32_outidx_i32_sv0_si0_rv1_maxtopk_1024_numthreads_512_occupancy_1_b_8192_b2_4096_tma_3.cu
Comment (DarkSharpness) [inline python/sglang/kernels/aot/csrc/deepselect/cuda_kernels/v3_fp32/instantiations/value_fp32_outidx_i32_sv0_si0_rv1_maxtopk_1024_numthreads_512_occupancy_1_b_8192_b2_4096_tma_3.cu:1]: Let's keep all the tuning in the official repo. We may seperate into `instantiations/sm90` and `instantiations/sm100`, and compile different variants for different arch.

**Review by DarkSharpness — CHANGES_REQUESTED** [review]

File: python/sglang/kernels/aot/python/sgl_kernel/__init__.py (hunk 2 above)
Comment (yuyu5333) [reply python/sglang/kernels/aot/python/sgl_kernel/__init__.py]: Thanks. Not restricted to SM90 anymore — addressed in the latest commits:
 - The build now emits sm_90a (CUDA ≥ 12.0) plus sm_100a / sm_103a (CUDA ≥ 12.9); cuobjdump confirms all three in the fatbin.
 - Dispatch selects the cluster variant per device at runtime (major == 9 → portable 8-CTA cluster, SM100/SM103 → 16-CTA). Device support is checked centrally in C++ (is_supported_device()), so the Python wrapper no longer hardcodes SM90.

File: python/sglang/kernels/aot/python/sgl_kernel/deepselect.py (hunk 3 above)
Comment (yuyu5333) [reply python/sglang/kernels/aot/python/sgl_kernel/deepselect.py]: Done — the wrapper now matches the official deep_select.topk():
 - Input accepts bf16 and fp32 (bf16 normal, bf16 cluster, and fp32 paths, each with int32/int64 indices).
 - Same signature and defaults as the official version: sorted, begin, end, indices_type, sorted_index, output_idx, output_idx_offset, idx_oob_fill_value, value_oob_fill_value, return_value, abort_when_nan_found.
 - On topk_lengths: the official interface spells this as end, where end[i] is the right-open endpoint of row i's valid range ([0, end[i])) — i.e. end[i] == topk_lengths[i] for varlen decode. We kept the official name rather than adding an alias so the two interfaces stay identical.

Comment (yuyu5333) [reply python/sglang/kernels/aot/python/sgl_kernel/deepselect.py]: In addition, passing a non-None begin/hint raises — the official kernel has no execution path for them (interface.py asserts the same). Varlen decode only needs end, which is fully supported, including the end[i] < topk gather-at-front + OOB-fill behavior.

File: python/sglang/kernels/aot/python/sgl_kernel/deepselect.py (hunk 4 above)
Comment (yuyu5333) [reply python/sglang/kernels/aot/python/sgl_kernel/deepselect.py]: Done.

File: python/sglang/kernels/aot/csrc/deepselect/cuda_kernels/v3_fp32/instantiations/value_fp32_outidx_i32_sv0_si0_rv1_maxtopk_1024_numthreads_512_occupancy_1_b_8192_b2_4096_tma_3.cu
Comment (yuyu5333) [reply python/sglang/kernels/aot/csrc/deepselect/cuda_kernels/v3_fp32/instantiations/value_fp32_outidx_i32_sv0_si0_rv1_maxtopk_1024_numthreads_512_occupancy_1_b_8192_b2_4096_tma_3.cu:1]: Done in the latest commit:
 - The arch-specific family is now split: v3_cluster/instantiations/sm90/ holds the portable 8-CTA cluster variants (dispatch only selects these on SM90), v3_cluster/instantiations/sm100/ the 16-CTA ones (SM100/SM103 only). v3 and v3_fp32 stay in shared directories — their tuning is workload-driven (num_waves, top-k tier), the same variant runs on every arch, so there are no per-arch variants to separate there.
 - The build now compiles each group only for arches that can reach it: shared instantiations for sm_90a/sm_100a/sm_103a, 8-CTA cluster only for sm_90a, 16-CTA cluster only for sm_100a+sm_103a.

File: python/sglang/srt/layers/attention/dsv4/candidate_indexer.py
Comment (yuyu5333) [reply python/sglang/srt/layers/attention/dsv4/candidate_indexer.py:1]: Thanks. Content other than AOT-kernel has been removed. I will submit a new PR to complete the integration of runtime code and kernel.
